topdown/sets.go

```diff
@@ -38,21 +38,49 @@ func builtinSetIntersection(a ast.Value) (ast.Value, error) {
 		return ast.NewSet(), nil
 	}
 
-	var result ast.Set
+	result := ast.NewSet()
+	var presentInAll map[*ast.Term]struct{}
+	keys := inputSet.Slice()
+
+	// Build index from first set.
+	first, err := builtins.SetOperand(keys[0].Value, 1)
+	if err != nil {
+		return nil, err
+	}
+	presentInAll = make(map[*ast.Term]struct{}, first.Len())
+	first.Iter(func(x *ast.Term) error {
+		presentInAll[x] = struct{}{}
+		return nil
+	})
 
+	// Keys slice MUST not change ordering before this, or we'll have issues.
+	isFirstItem := true
 	err = inputSet.Iter(func(x *ast.Term) error {
-		n, err := builtins.SetOperand(x.Value, 1)
+		// Skip first set in the list of sets.
+		if isFirstItem {
+			isFirstItem = false
+			return nil
+		}
+
+		curSet, err := builtins.SetOperand(x.Value, 1)
 		if err != nil {
 			return err
 		}
 
-		if result == nil {
-			result = n
-		} else {
-			result = result.Intersect(n)
+		// If a value isn't present in the current set, remove it from the list.
+		for k := range presentInAll {
+			if !curSet.Contains(k) {
+				delete(presentInAll, k)
+			}
 		}
+
 		return nil
 	})
+
+	// Add any surviving Terms to the output set.
+	for k := range presentInAll {
+		result.Add(k)
+	}
 	return result, err
```

Comment:
💭 `err`? Let's put an `if err != nil { ... }` right after the `err =` and `return result, nil` here